src/coreclr/System.Private.CoreLib/src/System/Variant.cs

```diff
@@ -78,323 +23,299 @@ internal partial struct Variant
         [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Variant_ConvertOleColorToSystemColor")]
         internal static partial void ConvertOleColorToSystemColor(ObjectHandleOnStack objret, uint value, IntPtr pMT);
 
-        //
-        // Native Methods
-        //
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal extern void SetFieldsObject(object val);
-
-        //
-        // Constructors
-        //
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Variant_ConvertValueTypeToRecord")]
+        private static partial void ConvertValueTypeToRecord(ObjectHandleOnStack obj, out ComVariant pOle);
 
-        internal Variant(int flags, object or, long data)
+        internal static ComVariant GetComIPFromObjectRef(object? obj)
         {
-            _flags = flags;
-            _objref = or;
-            _data = data;
+            IntPtr pUnk = GetIUnknownOrIDispatchForObject(ObjectHandleOnStack.Create(ref obj), out bool isIDispatch);
+            return ComVariant.CreateRaw(isIDispatch ? VarEnum.VT_DISPATCH : VarEnum.VT_UNKNOWN, pUnk);
         }
 
-        public Variant(bool val)
-        {
-            _objref = null;
-            _flags = CV_BOOLEAN;
-            _data = (val) ? bool.True : bool.False;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "MarshalNative_GetIUnknownOrIDispatchForObject")]
+        private static partial IntPtr GetIUnknownOrIDispatchForObject(ObjectHandleOnStack o, [MarshalAs(UnmanagedType.Bool)] out bool isIDispatch);
 
-        public Variant(sbyte val)
+        private static object? GetObjectRefFromComIP(IntPtr pUnk)
         {
-            _objref = null;
-            _flags = CV_I1;
-            _data = val;
+            return pUnk == IntPtr.Zero ? null : Marshal.GetObjectForIUnknown(pUnk);
         }
 
-        public Variant(byte val)
+        // Helper code for marshaling managed objects to VARIANT's
+        internal static void MarshalHelperConvertObjectToVariant(object? o, out ComVariant pOle)
         {
-            _objref = null;
-            _flags = CV_U1;
-            _data = val;
-        }
+            // Cases handled at native side: string, bool, primitives including (U)IntPtr excluding U(Int)64, array
 
-        public Variant(short val)
-        {
-            _objref = null;
-            _flags = CV_I2;
-            _data = val;
-        }
-
-        public Variant(ushort val)
-        {
-            _objref = null;
-            _flags = CV_U2;
-            _data = val;
-        }
+            pOle = default; // ComVariant.Create doesn't initialize all fields
```

Comment:
Managed test of S.R.InteropServices requires the fields to be initialized.
The cost of initialization should be very low comparing to managed-native transition. 128 bits is also a single SIMD instruction now.

Reply:
Well this does not work. The uninitialized field from `ComVariant.Create` will be copied and overwrite the initialized variable.
Should we change the behavior of `ComVariant.Create` to initialize it?

Reply:
@jkoritzinsky do you have any input here? I do want to reuse the logic in `ComVariant.Create`.

Reply:
Which test case is failing without fixing the initialization?

I'm open to changing `ComVariant.Create`, but I want to know the exact scenario before we do anything here.

Reply:
https://github.com/dotnet/runtime/blob/5fbe5c4c9b0c9fc80ab6de7614d5dc47aa4b7f7a/src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs#L114-L132

This test is asserting against the whole `IntPtr` value in VARIANT.

Reply:
Yeah there's no good way to do this differently and if we're depending on it in our tests, our customers definitely are. I think this is fine then.